### DISCUSSION TITLE
Aural Player Discussion Home

### DISCUSSION THREAD

**Discussion by kartik-venugopal** [kartik-venugopal/aural-player#76] [Announcements]
## 👋 Welcome!

<img width="225" src="https://raw.githubusercontent.com/maculateConception/aural-player/master/Documentation/Screenshots/readmeLogo.png"/>

 This is a great place to engage with the Aural Player project!
  * Ask questions you’re wondering about.
  * Share ideas, for example - suggest new features you'd like to see implemented.
  * Provide general feedback (either positive or negative)
  * If you're interested in somehow contributing to the project, let's talk!

**Comment by LesterJitsi** [comment]
Hi, glad to see you're bursting witht energy :wink: Aural is so good it's hard to suggest new features. I do have a little wish as unimportant as it may be. I'd like it was possible in _Menu bar mode_ to see the whole playlist currently playing in a drop-down menu, so that you can easily jump to the desired track instead of only moving to the previous or next (the playlist can be really long and so you might scroll up and down).  Maybe Ctrl-clicking Aural icon in the menu bar? No matter how.

I remember the old Vox (by @alenofx) had this handy feature among menu bar controls and I'm certain I already asked a developer to implement it, but I didn't manage to find out who he/she was. I just hope it wasn't you! Thanks.

**Reply by kartik-venugopal** [reply]
Hey @LesterJitsi , good to hear from you!

That is a very reasonable request for a feature that I think many will appreciate, and I will make sure to implement it in v4.0 (which I am working hard on and am determined to release in the next few months, come what may).

I will think of a good way to implement the play queue (formerly called the "playlist") in the menu bar.

Thanks again!

PS - I was not involved in the Vox project :)

**Reply by kartik-venugopal** [reply]
@LesterJitsi - It will look like this 😉 

![mb-pq](https://github.com/kartik-venugopal/aural-player/assets/30028539/0105a9bf-6461-4628-924e-c468b5d9ca69)

**Reply by alenofx** [reply]
Hi @LesterJitsi, thank you for mentioning me in this discussion.
Yes, I'm the original Vox developer, but didn't know about this aural-player;
it's seems an evolution of what the Vox project was at the time, congrats @kartik-venugopal, I'll check it out!

**Comment by LesterJitsi** [comment]
@alenofx, you've been my hero, I ran Vox from Tiger to Mojave, well after Coppertino turned it into an intrusive bloatware. Aural is a well worthy heir and more imho, basically the same philosophy :100:

**Comment by ixzh** [comment]
![Uploading image.png…]()
hey how to play searched for songs? like after `cmd+f`, the results are shown but not clickable.

**Reply by kartik-venugopal** [reply]
Hi, admittedly, v4 is very much in development, and some features may not work (I state this on the release page). But thanks anyway for letting me know.

But ... no play button showing on hover in the first column ?

**Reply by kartik-venugopal** [reply]
Hello, the tracks in search results are playable now.

**Comment by focusonsight** [comment]
Hey @kartik-venugopal, Aural is great! Thanks for all your hard work!

I have a question about a library feature. As I understand it, Aural is playlist based. You add a folder or files that you'd like to play. However, I'm seeking "library" functionality. I like to play my entire music collection on shuffle but Aural struggles with this because it takes a long time to load all the files (about 17,000 FLAC files). Once they are loaded everything works fine, but if you close and re-open Aural, it will take a long time to re-load all the files.

I'd like to be able to select music from inside Aural, instead of using finder to add files. This would be similar to Foobar's "media library" feature, where you point it to a directory and it monitors the directory for files and allows you to select a song, album, or entire library to play from within Foobar.

Is this possible with Aural?

**Reply by kartik-venugopal** [reply]
Hello @focusonsight, thanks for the nice feedback!. That's a great question, and the answer is yes! Aural v4 will come with a pretty full-featured library that will point to a folder and will monitor it for changes. I'm working really hard to release v4 later this year, and the Library is currently not complete, so the preview releases do not have the Library enabled. I'm sure that once I resume work on the Library, I will release a preview version that will let you play with the Library (before the final official release).

The Library will be organized into tracks, albums, artists, genres, decades, and playlist files. You will be able to select and play any of those groups easily with one click ... you could also play / shuffle the entire library if you so wish ... it will still take some time to load 17000 files (I have not yet tested the app with that many files), but playback should begin immediately.

Also keep in mind that even if you play something directly from the library, they will still be added to the Play Queue ... for Aural to keep track of what's playing, tracks must first be added to the Play Queue.

The new Library will also come with a "Tune Browser" file system view that is of course customized for music files / playlists.

See the 2 attached screenshots for a sneak peek at what it will look like.

Note - The screenshots do not show the custom playlists, history, favorites, and bookmarks that you will also be able to manage.

<img width="800" alt="Library-1" src="https://github.com/user-attachments/assets/b1afe9c7-2826-4811-9ea1-9fffd42e5d47">

<img width="800" alt="Library-2" src="https://github.com/user-attachments/assets/3910906b-1f38-43a8-8ce5-c2edfce7132c">

**Reply by focusonsight** [reply]
Awesome, great to hear! Thanks for the quick response. Do you know if the library feature will require all the tracks to load each time Aural is opened, or just for the first time initialization of the library? It seems with Foobar there is some cache of metadata because once the library is initialized and being monitored, you can close and re-open the app and instantly start viewing and playing anything in the library.

**Reply by kartik-venugopal** [reply]
I have implemented a metadata cache but that will likely be an option that can be enabled / disabled as some users may prefer to load the library fresh each time, especially if they like to mess around with tagging their files frequently, so it will be an option you can set. That should definitely help with library loading on app startup.

I really like that you mentioned your use case .. this is really valuable information because, to be honest, I don't know much about my app users - how large their libraries are, what file formats they prefer, etc. So thanks for that 👍  With information like what you mentioned, I can make sure my app is able to perform in such scenarios. I will try my best to test my library with 20,000 files and try to optimize for such cases.

Watch out later this year for the Library!

Cheers

**Reply by kartik-venugopal** [reply]
One idea - I could slightly modify the metadata cache right now to also work for tracks in the Play Queue ... that may help you at the moment, so that your 17000 files might load faster in the Play Queue upon startup. What do you think ?

And again, the cache would be optional (with a setting in Preferences).

**Comment by focusonsight** [comment]
Yes, the metadata option would definitely be nice to have! Just remembered I can also have the playlist clear when the app quits. That helps too, so when Aural restarts I don't have to wait for all the files to load if I just want to play 1 album.

Aural is the best music player I've found for Mac so it's strange that I haven't seen much chatter about it. Love the customization and FLAC support. People with local digital music (especially lossless) are sadly few and far between these days so the app support isn't great. I came from Foobar on Windows but unfortunately the Mac version is maybe 10% of what it is on Windows. 

The only other features on my wish list are visualizations, resizable album art panel, and resizing on the main panel. Although I don't really use effects, the audio units plugins feature is cool. I found the MiniMeters app which has great visualizations but unfortunately it's a standalone app and not available as an AU plugin (yet). Don't get me wrong, the built in visualizations are solid, they just leave me wanting more/missing the ones from Foobar. Let me know if you're aware of any good visualization plugins.

Really appreciate your work on this! Give me a shout if you need any help testing things!

**Reply by kartik-venugopal** [reply]
Again, thanks very much for the kind words 🙏 . I understand that Foobar is legendary in the music player world (Windows). My philosophy behind Aural is for it to be a decent Winamp equivalent on Mac that is easy, fun, and customizable. It will probably never be as powerful as IINA or some other players, but it will certainly evolve over time as I get feature requests and good feedback from app users. Because of time constraints (I am a one-man army), I sometimes have to weigh cost vs benefit when I get requests, but I will certainly read and consider requests.

I am currently working on ReplayGain ... once I release it (soon), I will implement the metadata caching option for the Play Queue, and then you can test it out and let me know if it helps you.

Visualizations - I don't have too many great ideas for what would make good visualizations (eg - shapes or visual patterns), but if you give me some ideas, I will definitely think about it! My initial idea was that I would provide a few visualizations out of the box and people can host AU plugins that would provide much better visualizations than I could ever create 😁  There are a few decent ones out there that are free. Take a look here: https://www.kvraudio.com/plugins/effects/windows/macosx/vst-plugins/vst3-plugins/audio-units/aax-plugins/rack-extensions/free/most-popular

Resizable album art panel - I assume you're referring to the Modular app mode ... it could be tricky, because all the info there is laid out in a very specific and constrained way ... in order to make room for a bigger art view, something else would have to be hidden or made smaller. Can you give more specific suggestions - What could be smaller and make room for the art to be bigger ?

Resizing on the main panel - If you're referring to the Modular app mode, I suppose it could be done (horizontally only), but the reason it has been left as is ... is that the main panel and Effects panel are meant to be the same exact size ... that makes it really easy to lay them out in a visually appealing way ... that was the original design. But again, if you could provide more details (do you mean horizontal resizing ?), I can consider this.

I appreciate your offer to test 👍  ... I will be putting out fairly frequent "preview" releases, so keep an eye out ... if you come across any bugs during your normal everyday use, feel free to file issues.

**Reply by focusonsight** [reply]
I had never heard of IINA but just tried it out. It's definitely polished but it lacks customization and it's pretty bare bones for a music player. I think it's going to be held back by trying to be a video **and** music player. Just a few examples: You can't make the playlist window wider (or really customize the views at all), so titles get cut off. No playlist manager, media library, or plugin support for lyrics, visualizations, etc. It's a good app but has a ways to go. I will say it loaded my whole FLAC library near instantly, not sure how it does that so fast.

Anyway, totally understand the need to look at cost/benefit for new Aural features. I'm sure my workflow/use case isn't the most common. I agree with your approach of providing a couple out of the box visualizations and leaving the rest to plugins. I do like the circular visualization in Aural but I prefer the standard music production/analysis visualizations like peak meter, waveform, oscilloscope, stereometer, spectrogram, and spectrum. Not sure how complicated it is to code those though. There's also a bunch of customizations that can go along with them. MiniMeters is quite good at this and has a free demo if you want to check it out. I think once it's available as an AU plugin I'll be good to go and won't really need built-in Aural visualizations. Most people probably don't care about visualizations but I find it interesting to _see_ the music as well as hear it 🙂

As for album art panel and resizing the main window, I'm referring to the windowed/modular mode (I do like the menu and control bar modes and use them from time to time). But with the windowed mode, yes, horizontal resizing would be nice, as well as vertical resizing. If I want to make the playlist window wider, I'm left with something like this:

![image](https://github.com/user-attachments/assets/8316fd91-9a47-4b23-855a-e9bf8e596827)

This assumes I'd be customizing the tracklist to show something like Artist - Album - Song, which would require a wider window. Generally I like to see that info in the tracklist, along with other info like year, bitrate, etc. Being able to resize the main window would open up that additional customization. If the main window is set to only show the track info, seek-bar, and playback buttons on mouse hover, then nothing should have to be hidden or made smaller. I don't really ever have the effects window open. Additionally, if I just have the main window open, I can't make it square to better fit the album art:

![image](https://github.com/user-attachments/assets/eff0f3fa-0247-4522-8f4d-bf7931086799)

A separate Aural panel for album art would be cool as well. I'm just used to the endless UI customization of Foobar 😄 All the UI components are fully interchangeable, resizable, etc.

Hope that info helps. I'll definitely keep an eye out for the preview builds! Thanks Kartik!

**Reply by kartik-venugopal** [reply]
Thanks for the info. I'm happy to report that my initial experimentation with the metadata caching for the Play Queue has produced decent results ... loading 3681 tracks has gone from about 6 - 7 seconds to about 0.4 seconds.

One thing I noticed - you are using v3.x ... I should tell you that I will not be doing any further work on v3 ... it is in archived status now. Any changes will be made only to v4.x.

Have you also tried the v4 preview releases ? There is now a waveform panel (recently added ... see screenshot). I can definitely add a peak / VU meter as I think many people could use that and I have implemented something like that before (different project).

![waveform-focusonsight](https://github.com/user-attachments/assets/1d420799-b1d5-4730-8baa-2c3bcd93e25a)

I can also do **_horizontal_** resizing of the main panel (it will likely retain the current width as a minimum), but I have to be honest - vertical resizing will complicate things a lot, so that idea will have to be shelved. 

A separate art panel is easy!

The v4 Play Queue has an alternate view that shows a bit more for each track, like the album name and art. Not sure if this helps in any way. I suppose I could make the track list a bit more customizable (what metadata you want to show ... bit rate, etc)


![PQ](https://github.com/user-attachments/assets/548c6858-a3bb-48a4-bc59-a604648e7051)

I will look at the pending work, the release timeline, estimated feature dev effort, and prioritize the work accordingly. For instance, the metadata cache is of the highest priority because it will affect every single user every time they use the app. Whereas a scrobbling integration for a platform no one uses may not be the highest prio.

I've filed the following issues for your feature requests.

https://github.com/kartik-venugopal/aural-player/issues/101
https://github.com/kartik-venugopal/aural-player/issues/102
https://github.com/kartik-venugopal/aural-player/issues/103
https://github.com/kartik-venugopal/aural-player/issues/104

**Reply by focusonsight** [reply]
Hey sorry for the slow reply, lots going on lately. Good to hear on the metadata caching! I just tried out the v4 preview, looks great! I'm liking the waveform and alternate view for the playlist. Looking forward to the separate art panel as well. If the main window can only be resized horizontally, no big deal. That would bridge the gap nicely when it comes to customizing the layout.

I only briefly tried v4 but the minimal view with art looks cool so I'll have to try it out along with the other features. The redesigned dock icon is a nice addition as well.

Totally understand the need to prioritize new features. I'm happy using Aural as it stands now so no worries on the stuff I've requested. Thanks for filing the issues for me! I'll circle back with you and/or file issues if anything comes up as I continue using the v4 preview.

**Reply by kartik-venugopal** [reply]
@focusonsight - Hey, I've released preview11 with metadata caching (can be controlled via preferences). Check it out. Also, the player window is horizontally resizable (not fully tested, some corner cases may be buggy, like auto-hiding of controls).

If you could please let me know how long it takes now to load your 17k files on startup, I'd really appreciate it! Thanks 😄

**Reply by kartik-venugopal** [reply]
@focusonsight - Also, horizontal resizing of player window released in preview13.

**Reply by focusonsight** [reply]
Hey Kartik! Just tried preview 17, looks like the metadata caching is working. All tracks are immediately loaded after closing and re-opening. It appeared to load my full flac library very quickly as well on the first try. Also, so far so good with the horizontal resizing. Thanks man!

One thing I noticed is that the previous and skip media keys don't work (play/pause works) when first opening Aural with the playback set to shuffle. When I turn off shuffle, the media keys work. Then I can re-enable shuffle and the media keys work again.

Lastly, if I'm shuffling my whole library and close Aural, am I able to have the playback resume from the same song/position? I enabled "Resume track playback from last playback position for all tracks" in the settings but when I re-opened Aural, it started playback from the top of the list. I also turned on autoplay.

**Reply by kartik-venugopal** [reply]
@focusonsight - Thanks for letting me know about all that! Nice to hear that the metadata caching is working !

---------------

EDIT - What looks like a media key bug is actually a bug with the shuffle logic. Good catch! You can still perform the step in the paragraph below, but it will not fix the previous / next issue ... that's a shuffle sequencing bug.

About the media keys - can you double check if you have added Aural in your System Settings > Privacy & Security > Accessibility ? Click the + button to add Aural.app, then toggle the little switch next to it, to allow Aural to control your computer. Right now, I think what's happening is that the play / pause is working via Control Center, not via the native media keys support. So try this and let me know what happens.

---------------

The last point is interesting! Right now, Aural does not remember your shuffle sequence between app launches. Every time you turn on Shuffle (or reorder tracks once Shuffle is ON), a new "shuffle sequence" is created which determines the order of playback. So, if you close the app, that sequence is lost ... then the next time you open the app and resume playback, you are creating a new shuffle sequence.

But, I can make Aural remember the shuffle sequence across app launches ... this is an interesting feature, haha 😄 I've created [issue #115 ](https://github.com/kartik-venugopal/aural-player/issues/115) to track this new work.

What Aural does currently have is under the History menu ... an option to "Resume last played track" .. this will work for normal sequential (non-shuffled) playback ... but I guess what you need is the same thing but when shuffle is on.

---------------

Lastly, about the "Resume track playback from last playback position for all tracks" .. what that does is ... on a per-track basis, it will remember where playback stopped for each individual track ... then the next time you play that track, it resumes from that position ... this is regardless of shuffle / repeat, etc. I think there are only very few cases when it will still start from the beginning when that option is enabled. 

I don't think this option is useful to you right now for playing the whole library. But you can use it when listening to a long audiobook / podcast file for instance ... let's say a 3 hour file, you listen to 1 hour, go away, come back, it starts from that 1 hour mark, and so on ...

**Reply by kartik-venugopal** [reply]
@focusonsight 

> One thing I noticed is that the previous and skip media keys don't work (play/pause works) when first opening Aural with the playback set to shuffle. When I turn off shuffle, the media keys work. Then I can re-enable shuffle and the media keys work again.
> 

Fixed in preview18

> Lastly, if I'm shuffling my whole library and close Aural, am I able to have the playback resume from the same song/position? I enabled "Resume track playback from last playback position for all tracks" in the settings but when I re-opened Aural, it started playback from the top of the list. I also turned on autoplay.

Now, you can either auto-resume (via Preferences > Playback > Autoplay > On app startup > Resume last played track sequence) or manually resume (via History > Resume last played track sequence) your shuffle sequence after restarting the app.

Available in preview20 ... just released ! Let me know if it works out well for you.

BTW, one more tip - there is a Play Queue preference to always load a folder on startup ... this is kind of like the library feature you wanted ... the real full-fledged library is also coming later, but for now, you could try this ... just point it to your music folder.

**Reply by focusonsight** [reply]
Awesome, thanks for fixing that! Looks like I stumbled on the play queue preference at some point because I had it set to my library folder, appreciate the heads up on that. 

Also, just tried the auto-resume, it works well! Love when I can just command+Q, go away, come back, and pick back up where I was in the shuffle sequence. Not sure if it's just in my head but it seems like sometimes with shuffling algos, the same songs have a tendency to pop up. In other words, if I want to hear every song in my library, it seems better to shuffle once and go through the whole playlist, than to shuffle again on each app startup 🤷‍♂️ 

Anyway, couple notes on preview 21. With the auto-resume, it takes 2-3 seconds for audio to play when the app is launched. When I was using Foobar, playback would start immediately when I open the app. Wondering if you have any ideas on that.

Second, when I try to skip a bunch of songs quickly with the media skip key on my keyboard, it causes Aural to hang. Does this have something to do with how many of the next shuffled songs are predetermined at a time?

Lastly, I just clicked the button to skip a song and Aural crashed. I attached the crash report here, hopefully it helps. Not sure I've ever had this happen so probably not a huge deal.

[Aural-p21-CrashReport.txt](https://github.com/user-attachments/files/18216026/Aural-p21-CrashReport.txt)

**Reply by kartik-venugopal** [reply]
Thanks for the feedback! You have really given me great insights into the app ... as the sole developer / user, I am not able to imagine how other people use the app, so thanks again!

So, ...

Shuffling algos - It is not super smart right now ... every shuffle sequence is completely new and does not take into account what was played before ... this is why the resume feature is helpful (if you care about history) ... if you just stop the player and then shuffle and play again, it will start a new sequence and recently played songs may pop up soon after because it just doesn't care what was played recently. But, now that you've mentioned this ... maybe I can make it a bit smarter ... avoid the recently played ones by looking at history. Good point 👍 

Auto-resume does take a few seconds to start, for a very good reason ... it must load the entire play queue so that the entire shuffle sequence can be re-constructed before it can start playing (which would be your 17000 tracks) ... I don't see a very good way around this. But I will keep it in mind and see if there is a quicker way to auto-resume.

About the slowness and crash ... I don't have any ideas and will have to dig deeper into your report (thanks for that!) and the code. It does not have anything to do with shuffling algos because the entire sequence is computed before playback starts. So, when you click next, it should simply pick the next one in the sequence and play it. Not sure why this is happening.

Would you be able to send me a screen recording of the slowness when you skip tracks quickly ? That might provide some insights. Thanks!

**Reply by kartik-venugopal** [reply]
PS - I will try shuffling all my 11K tracks and skip quickly to see if I can reproduce the issue(s).

**Reply by kartik-venugopal** [reply]
Lastly, I just clicked the button to skip a song and Aural crashed. I attached the crash report here, hopefully it helps. Not sure I've ever had this happen so probably not a huge deal.
> 
> [Aural-p21-CrashReport.txt](https://github.com/user-attachments/files/18216026/Aural-p21-CrashReport.txt)

I found the issue ... it's a very unlikely issue of a data structure being accessed simultaneously by multiple threads, and am fixing it. Thanks again for the crash report.

**Reply by focusonsight** [reply]
Awesome, thanks man! Glad to hear my nagging is appreciated 🤣 Give me a shout when it's ready and I'll test it out!

**Comment by bo1ta** [comment]
hey, I just came across this project and loving it!! please let me know if there are any bugs / open issues I could help with
